sycl/include/CL/sycl/INTEL/esimd/esimd_memory.hpp

```diff
@@ -238,6 +244,147 @@ ESIMD_INLINE ESIMD_NODEBUG void block_store(AccessorTy acc, uint32_t offset,
+/// @tparam T - element type; can only be a 1,2,4-byte integer or \c float,
+/// @tparam N  - the number of elements
+/// @tparam AccessorTy - \ref sycl::accessor type
+/// @tparam L1H - L1 cache hint
+/// @tparam L3H - L3 cache hint
+///
+/// Formal parameters:
+/// @param acc - the accessor to scatter to
+/// @param vals - values to write
+/// @param offsets - per-element offsets
+/// @param glob_offset - offset added to each individual element's offset to
+///   compute actual memory access offset for that element
+/// @param pred - per-element predicates; elements with zero corresponding
+///   predicates are not written
+///
+template <typename T, int N, typename AccessorTy,
+          CacheHint L1H = CacheHint::None, CacheHint L3H = CacheHint::None>
+ESIMD_INLINE ESIMD_NODEBUG
+    typename std::enable_if<(sizeof(T) <= 4) && (N == 1 || N == 8 || N == 16) &&
+                                !std::is_pointer<AccessorTy>::value,
+                            void>::type
+    scatter(AccessorTy acc, simd<T, N> vals, simd<uint32_t, N> offsets,
+            uint32_t glob_offset = 0, simd<uint16_t, N> pred = 1) {
+
+  constexpr int TypeSizeLog2 =
+      sycl::INTEL::gpu::ElemsPerAddrEncoding<sizeof(T)>();
+  // |TODO (performance) use hardware-supported scale once BE supports it
+  constexpr uint32_t scale = 0;
+  constexpr uint32_t t_scale = sizeof(T);
+  if constexpr (t_scale > 1) {
+    glob_offset *= t_scale;
+    offsets *= t_scale;
+  }
+
+  if constexpr (sizeof(T) < 4) {
+    static_assert(std::is_integral<T>::value,
+                  "only integral 1- & 2-byte types are supported");
+    using PromoT = typename std::conditional<std::is_signed<T>::value, int32_t,
+                                             uint32_t>::type;
+    const simd<PromoT, N> promo_vals = sycl::INTEL::gpu::convert<PromoT>(vals);
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__SYCL_EXPLICIT_SIMD__)
+    const auto surf_ind = AccessorPrivateProxy::getNativeImageObj(acc);
+    __esimd_surf_write<PromoT, N, decltype(surf_ind), TypeSizeLog2, L1H, L3H>(
+        pred, scale, surf_ind, glob_offset, offsets, promo_vals);
+#else
+    __esimd_surf_write<PromoT, N, AccessorTy, TypeSizeLog2, L1H, L3H>(
+        pred, scale, acc, glob_offset, offsets, promo_vals);
+#endif
+  } else {
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__SYCL_EXPLICIT_SIMD__)
+    const auto surf_ind = AccessorPrivateProxy::getNativeImageObj(acc);
+    __esimd_surf_write<T, N, decltype(surf_ind), TypeSizeLog2, L1H, L3H>(
+        pred, scale, surf_ind, glob_offset, offsets, vals);
+#else
+    __esimd_surf_write<T, N, AccessorTy, TypeSizeLog2, L1H, L3H>(
+        pred, scale, acc, glob_offset, offsets, vals);
+#endif
+  }
+}
+
+/// Load a scalar value from an accessor.
+template <typename T, typename AccessorTy, CacheHint L1H = CacheHint::None,
+          CacheHint L3H = CacheHint::None>
+ESIMD_INLINE ESIMD_NODEBUG T load(AccessorTy acc, uint32_t offset) {
```

Comment:
This is basically redundant to SIMD1 gather, which is not commonly used in CM for performance reason. Do we really need to expose a dedicated API for this? The naming is too general and could be a source of confusion, perhaps change to scalar_load to make a clear distinction with block_load.

Reply:
> This is basically redundant to SIMD1 gather

This is a convenient shortcut to calling generic gather/scatter. So I don't think it is redundant. `scalar_load` sounds good.

> is not commonly used in CM for performance reason. Do we really need to expose a dedicated API for this?

there are cases when scalar access is needed by algorithm (e.g. haar), so yes, I think we need this API for scalar access via accessors.

Reply:
OK. Please update name to scalar_load/scalar_store.

Reply:
ok